src/adapter/etl-adapter-csv/src/Flow/ETL/Adapter/CSV/CSVLineReader.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Flow\ETL\Adapter\CSV;
+
+use Flow\Filesystem\SourceStream;
+
+final readonly class CSVLineReader
+{
+    /**
+     * @param null|int<1, max> $charactersReadInLine
+     */
+    public function __construct(
+        private string $enclosure,
+        private ?int $charactersReadInLine = null,
+        private bool $removeBOM = true,
+    ) {
+    }
+
+    /**
+     * @return \Generator<int, string>
+     */
+    public function readLines(SourceStream $stream) : \Generator
+    {
+        $lineNumber = 0;
+        $buffer = '';
+
+        foreach ($stream->readLines(length: $this->charactersReadInLine) as $rawLine) {
+            $buffer .= $rawLine;
+
+            if (!\str_contains($buffer, $this->enclosure)) {
+                yield $this->removeBOM && $lineNumber === 0 ? $this->removeBOMFromLine(\rtrim($buffer, "\r\n")) : rtrim($buffer, "\r\n");
+                $lineNumber++;
+                $buffer = '';
+            } else {
+                if ($this->isCompleteCSVRecord($buffer)) {
+                    yield $this->removeBOM && $lineNumber === 0 ? $this->removeBOMFromLine(\rtrim($buffer, "\r\n")) : \rtrim($buffer, "\r\n");
+                    $lineNumber++;
+                    $buffer = '';
+                } else {
+                    $buffer .= "\n";
+                }
+            }
+        }
+
+        if ($buffer !== '') {
+            yield $this->removeBOM && $lineNumber === 0 ? $this->removeBOMFromLine(\rtrim($buffer, "\r\n")) : \rtrim($buffer, "\r\n");
```

Comment:
Something like this, probably:
```suggestion
        foreach ($stream->readLines(length: $this->charactersReadInLine) as $rawLine) {
            if ($this->removeBOM && $lineNumber === 0) {
                $rawLine = $this->removeBOMFromLine($rawLine);
            }

            $buffer .= $rawLine;

            if (!\str_contains($buffer, $this->enclosure)) {
                yield rtrim($buffer, "\r\n");
                $lineNumber++;
                $buffer = '';
            } else {
                if ($this->isCompleteCSVRecord($buffer)) {
                    yield \rtrim($buffer, "\r\n");
                    $lineNumber++;
                    $buffer = '';
                } else {
                    $buffer .= "\n";
                }
            }
        }

        if ($buffer !== '') {
            yield \rtrim($buffer, "\r\n");
```

Reply:
yeah but line number is not going to be incremented until it's full line, so at the end of the day if the first line contains multi line string, remove BOM will be executed as many times as many lines that multiline string has